File: symfony/src/Controller/Api/Content/ArticleController.php
<?php

namespace App\Controller\Api\Content;

use App\Entity\Content\Article;
use App\Repository\Content\ArticleRepository;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Attribute\Route;
use Symfony\Component\Validator\Validator\ValidatorInterface;
use App\Service\Content\ArticleService;
use App\Entity\Players\Player;
use App\Repository\Players\PlayerRepository;
use App\Entity\Cards\Deck;
use App\Repository\Cards\DeckRepository;

#[Route('/api/articles', name: 'article_')]
class ArticleController extends AbstractController
{
    public function __construct(
        private ArticleRepository $repository,
        private ValidatorInterface $validator,
        private ArticleService $service,
        private PlayerRepository $playerRepository,
        private DeckRepository $deckRepository,
    ) {}

    #[Route('', name: 'list', methods: ['GET'])]
    public function list(): JsonResponse
    {
        $items = $this->repository->findAll();
        return $this->json($items, context: ['groups' => ['article:read']]);
    }

    #[Route('', name: 'create', methods: ['POST'])]
    public function create(Request $request): JsonResponse
    {
        $data = json_decode($request->getContent(), true) ?? [];
        $article = new Article();
        if (isset($data['title'])) $article->setTitle($data['title']);
        if (isset($data['slug'])) $article->setSlug($data['slug']);
        if (isset($data['body'])) $article->setBody($data['body']);
        if (isset($data['excerpt'])) $article->setExcerpt($data['excerpt']);
        if (isset($data['coverImageUrl'])) $article->setCoverImageUrl($data['coverImageUrl']);
        if (isset($data['status'])) $article->setStatus($data['status']);
        if (isset($data['articleType'])) $article->setArticleType($data['articleType']);
        if (isset($data['language'])) $article->setLanguage($data['language']);
        if (isset($data['viewCount'])) $article->setViewCount($data['viewCount']);
        if (isset($data['likesCount'])) $article->setLikesCount($data['likesCount']);
        if (isset($data['isFeatured'])) $article->setIsFeatured($data['isFeatured']);
        if (isset($data['publishedAt'])) $article->setPublishedAt(new \DateTime($data['publishedAt']));
        if (isset($data['createdAt'])) $article->setCreatedAt(new \DateTime($data['createdAt']));
        if (isset($data['updatedAt'])) $article->setUpdatedAt(new \DateTime($data['updatedAt']));
        if (!isset($data['author'])) return $this->json(['error' => 'author is required'], Response::HTTP_UNPROCESSABLE_ENTITY);
        $rel_author = $this->playerRepository->find($data['author']);
        if (!$rel_author) return $this->json(['error' => 'Player not found'], Response::HTTP_UNPROCESSABLE_ENTITY);
        $article->setAuthor($rel_author);
        if (array_key_exists('featuredDeck', $data)) {
            $article->setFeaturedDeck($data['featuredDeck'] !== null ? $this->deckRepository->find($data['featuredDeck']) : null);
        }

        $errors = $this->validator->validate($article);
        if (count($errors) > 0) {
            return $this->json(['errors' => (string) $errors], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        try {
            $article->validateImplies();
        } catch (\DomainException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        $this->repository->save($article, flush: true);
        return $this->json($article, Response::HTTP_CREATED, context: ['groups' => ['article:read']]);
    }

    #[Route('/{id}', name: 'show', methods: ['GET'])]
    public function show(Article $article): JsonResponse
    {
        return $this->json($article, context: ['groups' => ['article:read']]);
    }

    #[Route('/{id}', name: 'update', methods: ['PUT', 'PATCH'])]
    public function update(Request $request, Article $article): JsonResponse
    {
        $data = json_decode($request->getContent(), true) ?? [];
        if (isset($data['title'])) $article->setTitle($data['title']);
        if (isset($data['slug'])) $article->setSlug($data['slug']);
        if (isset($data['body'])) $article->setBody($data['body']);
        if (isset($data['excerpt'])) $article->setExcerpt($data['excerpt']);
        if (isset($data['coverImageUrl'])) $article->setCoverImageUrl($data['coverImageUrl']);
        if (isset($data['status'])) $article->setStatus($data['status']);
        if (isset($data['articleType'])) $article->setArticleType($data['articleType']);
        if (isset($data['language'])) $article->setLanguage($data['language']);
        if (isset($data['viewCount'])) $article->setViewCount($data['viewCount']);
        if (isset($data['likesCount'])) $article->setLikesCount($data['likesCount']);
        if (isset($data['isFeatured'])) $article->setIsFeatured($data['isFeatured']);
        if (isset($data['publishedAt'])) $article->setPublishedAt(new \DateTime($data['publishedAt']));
        if (isset($data['createdAt'])) $article->setCreatedAt(new \DateTime($data['createdAt']));
        if (isset($data['updatedAt'])) $article->setUpdatedAt(new \DateTime($data['updatedAt']));
        if (isset($data['author'])) {
            $rel_author = $this->playerRepository->find($data['author']);
            if (!$rel_author) return $this->json(['error' => 'Player not found'], Response::HTTP_UNPROCESSABLE_ENTITY);
            $article->setAuthor($rel_author);
        }
        if (array_key_exists('featuredDeck', $data)) {
            $article->setFeaturedDeck($data['featuredDeck'] !== null ? $this->deckRepository->find($data['featuredDeck']) : null);
        }

        $errors = $this->validator->validate($article);
        if (count($errors) > 0) {
            return $this->json(['errors' => (string) $errors], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        try {
            $article->validateImplies();
        } catch (\DomainException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }

        $this->repository->save($article, flush: true);
        return $this->json($article, context: ['groups' => ['article:read']]);
    }

    #[Route('/{id}', name: 'delete', methods: ['DELETE'])]
    public function delete(Article $article): JsonResponse
    {
        $this->repository->remove($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/publish', name: 'publish', methods: ['POST'])]
    public function publish(Article $article): JsonResponse
    {
        $article->publish();
        $this->repository->save($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/archive', name: 'archive', methods: ['POST'])]
    public function archive(Article $article): JsonResponse
    {
        $article->archive();
        $this->repository->save($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/view', name: 'incrementView', methods: ['POST'])]
    public function incrementView(Article $article): JsonResponse
    {
        $article->incrementView();
        $this->repository->save($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/like', name: 'like', methods: ['POST'])]
    public function like(Article $article): JsonResponse
    {
        $article->like();
        $this->repository->save($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/like', name: 'unlike', methods: ['DELETE'])]
    public function unlike(Article $article): JsonResponse
    {
        $article->unlike();
        $this->repository->save($article, flush: true);
        return $this->json(null, Response::HTTP_NO_CONTENT);
    }

    #[Route('/{id}/reading-time', name: 'readingTimeMinutes', methods: ['GET'])]
    public function readingTimeMinutes(Article $article): JsonResponse
    {
        $result = $article->readingTimeMinutes();
        $this->repository->save($article, flush: true);
        return $this->json($result);
    }
    #[Route('/{id}/transitions/draft-to-published', name: 'article_transitionDraftToPublished', methods: ['PATCH'])]
    public function transitionDraftToPublished(Article $article): JsonResponse
    {
        try {
            $result = $this->service->transitionDraftToPublished($article->getId());
            return $this->json($result);
        } catch (\Symfony\Component\HttpKernel\Exception\ConflictHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_CONFLICT);
        } catch (\Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }
    }

    #[Route('/{id}/transitions/published-to-archived', name: 'article_transitionPublishedToArchived', methods: ['PATCH'])]
    public function transitionPublishedToArchived(Article $article): JsonResponse
    {
        try {
            $result = $this->service->transitionPublishedToArchived($article->getId());
            return $this->json($result);
        } catch (\Symfony\Component\HttpKernel\Exception\ConflictHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_CONFLICT);
        } catch (\Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }
    }

    #[Route('/{id}/transitions/archived-to-draft', name: 'article_transitionArchivedToDraft', methods: ['PATCH'])]
    public function transitionArchivedToDraft(Article $article): JsonResponse
    {
        try {
            $result = $this->service->transitionArchivedToDraft($article->getId());
            return $this->json($result);
        } catch (\Symfony\Component\HttpKernel\Exception\ConflictHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_CONFLICT);
        } catch (\Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException $e) {
            return $this->json(['error' => $e->getMessage()], Response::HTTP_UNPROCESSABLE_ENTITY);
        }
    }

    #[Route('/{id}/transitions/published-to-draft', name: 'article_transitionPublishedToDraft', methods: ['PATCH'])]
    public function transitionPublishedToDraft(Article $article): JsonResponse
    {
        return $this->json(['error' => 'Transition Published -> Draft is not allowed'], Response::HTTP_CONFLICT);
    }
}

File: phoenix/lib/cards_project_web/controllers/content/article_controller.ex
defmodule CardsProjectWeb.Content.ArticleController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Content
  alias CardsProject.Content.Article

  def index(conn, params) do
    q = Map.get(params, "q")
    articles = Content.list_articles(q)
    json(conn, Enum.map(articles, &serialize_article/1))
  end

  def show(conn, %{"id" => id}) do
    article = Content.get_article!(id)
    json(conn, serialize_article(article))
  end

  def create(conn, params) do
    case Content.create_article(params) do
      {:ok, article} ->
        conn
        |> put_status(:created)
        |> json(serialize_article(article))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def update(conn, %{"id" => id} = params) do
    article = Content.get_article!(id)
    case Content.update_article(article, params) do
      {:ok, article} ->
        json(conn, serialize_article(article))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  # POST /api/articles/{id}/publish
  def publish(conn, %{"id" => id}) do
    Content.article_publish_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/articles/{id}/archive
  def archive(conn, %{"id" => id}) do
    Content.article_archive_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/articles/{id}/view
  def increment_view(conn, %{"id" => id}) do
    Content.article_increment_view_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/articles/{id}/like
  def like(conn, %{"id" => id}) do
    Content.article_like_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # DELETE /api/articles/{id}/like
  def unlike(conn, %{"id" => id}) do
    Content.article_unlike_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # GET /api/articles/{id}/reading-time
  def reading_time_minutes(conn, %{"id" => id}) do
    result = Content.article_reading_time_minutes_behavior(id)
    json(conn, %{result: result})
  end

  # PATCH /api/articles/:id/transitions/draft-to-published
  def transition_draft_to_published(conn, %{"id" => id}) do
    article = Content.get_article!(id)
    case Content.transition_draft_to_published_article(article) do
      {:ok, updated} ->
        json(conn, serialize_article(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/articles/:id/transitions/published-to-archived
  def transition_published_to_archived(conn, %{"id" => id}) do
    article = Content.get_article!(id)
    case Content.transition_published_to_archived_article(article) do
      {:ok, updated} ->
        json(conn, serialize_article(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/articles/:id/transitions/archived-to-draft
  def transition_archived_to_draft(conn, %{"id" => id}) do
    article = Content.get_article!(id)
    case Content.transition_archived_to_draft_article(article) do
      {:ok, updated} ->
        json(conn, serialize_article(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/articles/:id/transitions/published-to-draft
  def transition_published_to_draft(conn, %{"id" => _id}) do
    conn
    |> put_status(:conflict)
    |> json(%{error: "Transition Published -> Draft is not allowed"})
  end

  defp serialize_article(%Article{} = record) do
    record
    |> Map.take([:id, :title, :slug, :body, :excerpt, :cover_image_url, :status, :article_type, :language, :view_count, :likes_count, :is_featured, :published_at, :created_at, :updated_at, :author_id, :featured_deck_id, :comments_id])
    |> (fn m -> Map.put(Map.delete(m, :created_at), :created_at, Map.get(m, :created_at)) end).()
    |> (fn m -> Map.put(Map.delete(m, :updated_at), :updated_at, Map.get(m, :updated_at)) end).()
    |> (fn m -> Map.put(Map.delete(m, :published_at), :published_at, Map.get(m, :published_at)) end).()
  end

  defp format_errors(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end
end
